Subject: [PATCH] PackageManager: Init known locales at WelcomeActivity in SetupWizard

Because of ShortcutUser.mKnownLocales not initialized yet, so ShortcutService do nothing when
handling LOCALE_CHANGED broadcast.

Test: make services and do factory reset. All apps can show correct language resource strings
after changing language in WelcomActivity UI of SetupWizard.

diff --git a/services/core/java/com/android/server/pm/ShortcutService.java b/services/core/java/com/android/server/pm/ShortcutService.java
index 15d20716a32..d179d7e48e3 100644
--- a/services/core/java/com/android/server/pm/ShortcutService.java
+++ b/services/core/java/com/android/server/pm/ShortcutService.java
@@ -1181,6 +1181,8 @@ ShortcutUser getUserShortcutsLocked(@UserIdInt int userId) {
             if (userPackages == null) {
                 userPackages = new ShortcutUser(this, userId);
             }
+
+            userPackages.initKnownLocales();
             mUsers.put(userId, userPackages);
 
             // Also when a user's data is first accessed, scan all packages.
diff --git a/services/core/java/com/android/server/pm/ShortcutUser.java b/services/core/java/com/android/server/pm/ShortcutUser.java
index 2c388c4a44d..ef66adb74be 100644
--- a/services/core/java/com/android/server/pm/ShortcutUser.java
+++ b/services/core/java/com/android/server/pm/ShortcutUser.java
@@ -286,6 +286,10 @@ private String getKnownLocales() {
         return mKnownLocales;
     }
 
+    public void initKnownLocales(){
+        getKnownLocales();
+    }
+
     /**
      * Check to see if the system locale has changed, and if so, reset throttling
      * and update resource strings.
@@ -296,9 +300,10 @@ public void detectLocaleChange() {
             return;
         }
         if (ShortcutService.DEBUG) {
-            Slog.d(TAG, "Locale changed from " + currentLocales + " to " + mKnownLocales
+            Slog.d(TAG, "Locale changed from " + mKnownLocales + " to " + currentLocales
                     + " for user " + mUserId);
         }
+
         mKnownLocales = currentLocales;
 
         forAllPackages(pkg -> {
